### ISSUE TITLE
Where are these?

### ISSUE THREAD

**Issue by GitAlasGo** [Julow/Unexpected-Keyboard#801]
This is the keys, that i want showing up in the arabic keyboard: https://imgur.com/a/jhd5uKT

I could add them in the custom keys, but the problem is that they're not noticeable.

Maybe i'm missing something obvious?

---

Also i'm a beginner in arabic, so i won't be able to read any arabic replies.

**Comment by Julow** [comment]
These are missing from the keyboard entirely. Do you have the unicode number for all of these ?

**Comment by GitAlasGo** [comment]
> These are missing from the keyboard entirely. Do you have the unicode number for all of these ?

Okay here's all the unicode, Github have a copy button on the right side, so you can click on it to copy them.

ARABIC FATHA, U+064E
```
َ
```
ARABIC KASRA, U+0650
```
ِ
```
ARABIC DAMMA, U+064F
```
ُ
```
ARABIC MADDAH ABOVE, U+0653
```
ٓ
```
ARABIC LETTER SUPERSCRIPT ALEF, U+0670
```
ٰ
```
ARABIC SUBSCRIPT ALEF, U+0656
```
ٖ
```
ARABIC SUKUN, U+0652
```
ْ
```
ARABIC FATHATAN, U+064B
```
ً
```
ARABIC KASRATAN, U+064D
```
ٍ
```
ARABIC DAMMATAN, U+064C
```
ٌ
```
ARABIC SHADDA, U+0651
```
ّ
```
ARABIC HAMZA ABOVE, U+0654
```
ٔ
```
ARABIC HAMZA BELOW, U+0655
```
ٕ
```

**Comment by Julow** [comment]
Done in https://github.com/Julow/Unexpected-Keyboard/pull/807. Can you try the debug build at the end of [this page](https://github.com/Julow/Unexpected-Keyboard/actions/runs/11986915241?pr=807) to see if they are all working correctly ?

**Comment by GitAlasGo** [comment]
> Done in #807. Can you try the debug build at the end of [this page](https://github.com/Julow/Unexpected-Keyboard/actions/runs/11986915241?pr=807) to see if they are all working correctly ?

It's working, and there's finally a circle under them.

![Screenshot_20241123-150526](https://github.com/user-attachments/assets/99068d04-b2e9-451f-8874-9f198a202106)

That little circle doesn't appear, if i add them to the **Custom keys**

**Comment by GitAlasGo** [comment]
@Julow How do i make the **Arabic diacritics** only show up in arabic layouts?

It's showing the diacritics on every other layout like english, the **Arabic diacritics** are basically useless in english layout.

Picture of **Arabic diacritics** showing in english layout
![Screenshot_20241123-215222](https://github.com/user-attachments/assets/e7bdc573-1914-4a29-8c71-0d47c4578c10)

**Comment by Julow** [comment]
Good catch! To fix this bug, the easiest way would be to add the diacritics into the layouts directly and not in the settings. Do you think you could update the layouts ? The layouts are here: https://github.com/Julow/Unexpected-Keyboard/tree/master/srcs/layouts

**Comment by GitAlasGo** [comment]
> Good catch! To fix this bug, the easiest way would be to add the diacritics into the layouts directly and not in the settings. Do you think you could update the layouts ? The layouts are here: https://github.com/Julow/Unexpected-Keyboard/tree/master/srcs/layouts

@Julow Here is the layout with the arabic diacritics included
```
<?xml version="1.0" encoding="utf-8"?>
<keyboard name="Arabic Alt" script="arabic" numpad_script="hindu-arabic">
  <row>
    <key key0="ض" key2="١" key3="`" key4="loc esc"/>
    <key key0="ص" key2="٢" key3="\@"/>
    <key key0="ث" key2="٣" key3="\#"/>
    <key key0="ق" key2="٤" key3="$"/>
    <key key0="ف" key2="٥" key3="%"/>
    <key key0="غ" key2="٦" key3="^"/>
    <key key0="ع" key2="٧" key3="&amp;"/>
    <key key0="ه" key2="٨" key3="*"/>
    <key key0="خ" key2="٩" key3="("/>
    <key key0="ح" key2="٠" key3=")"/>
    <key key0="ج"/>
  </row>
  <row>
    <key key0="ش" key3="combining_dammatan" key4="loc tab"/>
    <key key0="س" key3="combining_alef_below" key4="combining_alef_above"/>
    <key key0="ي" key4="combining_fatha"/>
    <key key0="ب" key4="combining_shaddah"/>
    <key key0="ل" key4="combining_dammah"/>
    <key key0="ا" key1="أ" key4="combining_sukun"/>
    <key key0="ت" key4="combining_fathatan"/>
    <key key0="ن" key4="combining_kasra"/>
    <key key0="م" key4="combining_kasratan"/>
    <key key0="ك" key4="combining_hamza_above"/>
    <key key0="ط" key4="combining_hamza_below"/>
  </row>
  <row>
    <key key0="ذ"/>
    <key key0="ء"/>
    <key key0="ؤ" key1="{"/>
    <key key0="ر" key1="}"/>
    <key key0="ى" key1="ئ"/>
    <key key0="ة"/>
    <key key0="و" key3=","/>
    <key key0="ز" key3="."/>
    <key key0="ظ" key3="&#1567;"/>
    <key key0="د"/>
    <key width="1.0" key0="backspace" key2="delete"/>
  </row>
</keyboard>
```

**Comment by Julow** [comment]
Thanks :) In https://github.com/Julow/Unexpected-Keyboard/pull/807 I'm adding `Fn` combos for some of these with @anaskaejdar 
To avoid cluttering the layout with all the diacritics, do you think more `Fn` combos could be added ?

**Comment by GitAlasGo** [comment]
> Thanks :) In #807 I'm adding `Fn` combos for some of these with @anaskaejdar To avoid cluttering the layout with all the diacritics, do you think more `Fn` combos could be added ?

@Julow I'm struggling to understand, where to add more letters to the **fn**

I went to check the commit and i saw `arabic.json`, `fn.json` i forked the repo, since i'm on a phone, and github can compile them for me.

I changed a few things in these files, but when i test them i don't see any changes.

I saw a few diacritics in the **FN**, but not all the diacritics are there.

---

Sorry if this sounds dumb, i'm not a programmer.

**Comment by anaskaejdar** [comment]
Yes my friend, the FN part simply swaps the diacritics for their full-letter counterparts and vise versa, so if you already have ێ on the keyboard, FN will turn it into its diacritic equivalent, kasra, and if you already have kasra on your keyboard, FN will turn it into ێ

But if you have neither, it's not Fn you want, but rather the Compose key (right next to the enter key on the bottom right)

Press compose, then press ي twice to get kasra
Press compose, then press ه twice to get fatha
Press compose, then press و twice to get damma
Press compose, then press ا twice to get small-alef-above
Press compose, then press ح twice to get sukun

And for the nasalized versions, press the same letter only once, and then press ن

Notable exceptions (for arabic):
Fn+ة will also get you the fatha
Fn+ئ will also get you hamze-above

Also, there are some other diacritics you can get by pressing Compose, then pressing either ۷ or ۸; for instance, small-alef-below, hamze-below, or the V-shaped diacritics

**Comment by Julow** [comment]
@GitAlasGo `fn.json` defines the Fn mappings and `arabic.json` defines Compose sequences but modifying these files is not enough to add more things to Fn and Compose unfortunately.
You must run `gradle test` to update the app's code.
You didn't sound dumb at all, the code is dumb for being unintuitive.

@anaskaejdar Thanks you very much for these well thought Fn and Compose mappings.
I just want to add that Compose sequences are not intended for commonly used symbols because they are hard to discover and unconventional. It would be best if all the diacritics or other letters that people need every day are on the Fn layer.

**Comment by anaskaejdar** [comment]
@Julow Hey happy thanksgiving!

Yeah, I'm down to add the rest of the diacritics into the FN layer such that it would appear in the arabic keyboard. My idea was just to make these symbols accessible, because a lot of languages that use this alphabet don't really use the diacritics except in rare circumstances to disambiguate homographs if needed.

I assume that if a language has a direct need for them, like arabic itself does, they would be directly encoded into the keyboard for that language, and the compose sequences would be more useful for accessing letters that belong to other arabic-alphabet languages.

From the perspective of persian or kurdish, the diacritics themselves are essentially foreign letters.

But nonetheless, I'm down to work on this. Good news is there's a bunch of letters that still don't yet have any FN alternate assigned, so there's room to work there.

**Comment by Julow** [comment]
@GitAlasGo You can find the latest debug build at the end of this page if you want to try out the current Fn and Compose combos: https://github.com/Julow/Unexpected-Keyboard/actions/runs/11995191558

@anaskaejdar Thanks for your help :) Let's see which diacritics are actually needed.

**Comment by Julow** [comment]
Sorry, the correct link is https://github.com/Julow/Unexpected-Keyboard/actions/runs/12169833293
